### PR TITLE
Implement audit findings

### PR DESCRIPTION
**Pull request by paluh**
#### 2.3.1 Unexpected behavior

2.3.1.1 Full value conservation is too strong a requirement:
* Implementation: https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/src/CredentialManager/Scripts/Common.hs#L155
* Tests:
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/test/CredentialManager/Gen.hs#L442
  * All `Value` checks were updated - I list one two examples: one for `HotNFT` and one for `ColdNFT`:
    * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/test/CredentialManager/Scripts/ColdNFT/AuthorizeSpec.hs#L221
    * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/test/CredentialManager/Scripts/HotNFT/ResignVotingSpec.hs#L224
  
2.3.1.2 Misleading burning check in spending script:
* Implementation: https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/src/CredentialManager/Scripts/Common.hs#L231
* Tests:
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/test/CredentialManager/Scripts/ColdNFT/BurnColdSpec.hs#L107
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/test/CredentialManager/Scripts/HotNFT/BurnHotSpec.hs#L167
  
2.3.1.3 Upgrading can be done to manipulate datum in the same script
* Implementation: https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/src/CredentialManager/Scripts/Common.hs#L242
* Tests:
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/test/CredentialManager/Scripts/ColdNFT/UpgradeColdSpec.hs#L146
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/test/CredentialManager/Scripts/HotNFT/UpgradeHotSpec.hs#L207
  
2.3.1.4 NFT Script does not check for NFT presence
* Implementation: https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/src/CredentialManager/Scripts/Common.hs#L108
* Tests:
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/test/CredentialManager/Scripts/ColdNFTSpec.hs#L116
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/test/CredentialManager/Scripts/HotNFTSpec.hs#L108
  * Many scenarios were failing and all of them were updated so their corresponding `ValidArgs` provide script input with appropriate - an example:
    * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/test/CredentialManager/Scripts/ColdNFT/AuthorizeSpec.hs#L390
    
2.3.1.5 `findOutputsByCredential` could filter by address
* Implementation: https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/src/CredentialManager/Scripts/Common.hs#L150

#### 2.3.2 Unclear specification

2.3.2.1 Unchecked reference script in `checkContinuingTx`
* Implementation:
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/src/CredentialManager/Scripts/Minting.hs#L150
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/src/CredentialManager/Scripts/Common.hs#L152
* Tests:
  * Valid `ScriptArgs` generation was changed in both `ColdNFTSpec.hs` and `HotNFTSpec.hs`:
    * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/test/CredentialManager/Scripts/ColdNFTSpec.hs#L529
    * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/test/CredentialManager/Scripts/HotNFTSpec.hs#L491
  * Extra tests were added:
    * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/test/CredentialManager/Scripts/HotNFTSpec.hs#L177
    * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/test/CredentialManager/Scripts/ColdNFTSpec.hs#L186
    
2.3.2.2 NFT existence checks are too liberal
* Implementation - all corresponding checks `> 0` (or `0 <`) where turned into `== 1` - examples:
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/src/CredentialManager/Scripts/Common.hs#L108
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/src/CredentialManager/Scripts/Common.hs#L246
  
  
2.3.2.3 `checkUpgrade` function does not match its specification
* Implementation - specification was updated accordingly:
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/doc/read-the-docs-site/system-overview/cold-nft.rst#rules-for-upgradecold-actions
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/doc/read-the-docs-site/system-overview/hot-nft.rst#rules-for-upgradehot-actions

2.3.2.4 Datum upgrade requirements undocumented
* Implementation - warning section was added under `UpgradeCold` rules section:
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/doc/read-the-docs-site/system-overview/cold-nft.rst#rules-for-upgradecold-actions
  
2.3.2.5
* Implementation - inline datum requirement was added to the scripts rules lists:
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/doc/read-the-docs-site/system-overview/hot-nft.rst#rules
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/doc/read-the-docs-site/system-overview/cold-nft.rst#rules

#### 2.3.3 Code quality

2.3.3.1 Double instance of a `checkBurn` function
* Implementation:
  * `Common.checkBurn` was renamed to `Common.checkBurnAll` which checks if all tokens were burned: https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/src/CredentialManager/Scripts/Common.hs#L228
  * `Mining.checkBurn` was split into two checks:
    * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/src/CredentialManager/Scripts/Minting.hs#L106
    * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/src/CredentialManager/Scripts/Minting.hs#L112
    
2.3.3.2 `checkSpendingTx` re-implements Plutus logic
* Implementation - switched to `findOwnInput`:
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/src/CredentialManager/Scripts/Common.hs#L101
  
2.3.3.3 Unnecessary `Datum` dependency check
* Implementation - all non datum dependent checks where extracted out:
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/src/CredentialManager/Scripts/ColdNFT.hs#L40

2.3.3.4 Argument names for `checkMultiSig` are unclear
* Implemention:
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/src/CredentialManager/Scripts/Common.hs#L68
  
2.3.3.5 Unused parameter to `checkSpendingTx` callback
* Implementation:
  * https://github.com/IntersectMBO/credential-manager/blob/paluh/audit-findings/credential-manager/src/CredentialManager/Scripts/Common.hs#L88

2.3.3.6 Use of `RecordWildCards` extension
* Implementation - switched to field puns in all the on-chain code.

### CODE REVIEW

**Review by perturbing — COMMENTED** [review]
Great work @paluh, truly grateful for the work that you and @jhbertra did to lift this project to the next level. The code base is very well written now, and I am proud of how the solution turned out and learned a ton from you, thanks again both!

Regarding the implementation of the second round audit, I have a few questions. I think that in general the correct changes are made that are suggested in the report. I want to be complete in my feedback, so I am a bit explicit. I have the following comments:

**Unexpected behavior**
- 2.3.1.1: This PR correctly resolves the findings.
- 2.3.1.2: This PR correctly resolves the findings.
- 2.3.1.3: This PR correctly resolves the findings. Additionally, I would like to mention that this was not per se a bad thing, since ultimately, the member group has full control. I think this may be a feature, for example it could be a quick fix in the case of an unexpected corrupt datum. I am confident that this will not be the case, and perhaps this is a case of the idiom potato or potato, since in that case, the update action is still present. What do you think?
- 2.3.1.4: This PR correctly resolves the findings.
- 2.3.1.5: This PR correctly resolves the findings, does our off chain still use no staking credential part? Or do we allow for one now?

**Unclear specification**
- 2.3.2.1: This PR correctly resolves the findings.
- 2.3.2.2: This PR correctly resolves the findings.
- 2.3.2.3: The only change you linked it the specification, am I correct that the code that ensures this, is [this](https://github.com/IntersectMBO/credential-manager/blob/99a66b22f6183041e8c268957b7801340d5777e7/credential-manager/src/CredentialManager/Scripts/Common.hs#L108) code?
- 2.3.2.4: I think we should add more visably warning text formatting for the warnings. Something like  ⚠️ WARNING ?
- 2.3.2.5: audit also specified better / more verbose error handling, is this also implemented?

**Code quality**
- 2.3.3.1: This PR correctly resolves the findings.
- 2.3.3.2: This PR correctly resolves the findings.
- 2.3.3.3: This PR correctly resolves the findings.
- 2.3.3.4: This PR correctly resolves the findings. What do you think about adding an example table to the specification that exampliefies what majority means exactly.
- 2.3.3.5: This PR correctly resolves the findings.
- 2.3.3.6: This PR correctly resolves the findings.

Again, great pr :+1:

File: credential-manager/src/CredentialManager/Scripts/ColdCommittee.hs [hunk 1]
```diff
@@ -29,13 +29,13 @@ import PlutusTx.Prelude hiding (trace, traceIfFalse)
 -- in any spending input of the transaction.
 {-# INLINEABLE coldCommitteeScript #-}
 coldCommitteeScript :: AssetClass -> ScriptContext -> Bool
-coldCommitteeScript nft ScriptContext{..} = case scriptContextScriptInfo of
+coldCommitteeScript nft ScriptContext{scriptContextScriptInfo, scriptContextTxInfo} = case scriptContextScriptInfo of
```
Comment (jhbertra) [inline credential-manager/src/CredentialManager/Scripts/ColdCommittee.hs:32]: 2.3.3.6

File: credential-manager/src/CredentialManager/Scripts/ColdCommittee.hs [hunk 2]
```diff
@@ -29,13 +29,13 @@ import PlutusTx.Prelude hiding (trace, traceIfFalse)
...
   CertifyingScript _ _ ->
     traceIfFalse "Cold NFT not found in any input" $ any inputSpendsToken txInputs
   _ -> trace "Invalid script purpose" False
   where
     -- Checks if an input spends the correct token
-    inputSpendsToken TxInInfo{txInInfoResolved = TxOut{..}} =
+    inputSpendsToken TxInInfo{txInInfoResolved = TxOut{txOutValue}} =
```
Comment (jhbertra) [inline credential-manager/src/CredentialManager/Scripts/ColdCommittee.hs:38]: 2.3.3.6

File: credential-manager/src/CredentialManager/Scripts/ColdNFT.hs [hunk 3]
```diff
@@ -37,42 +37,42 @@ coldNFTScript
...
-  checkSpendingTx \TxInfo{..} _ inAddress inValue datumIn -> \case
+  checkSpendingTx coldNFT \txInfo@TxInfo{txInfoTxCerts, txInfoOutputs, txInfoSignatories} inAddress inValue inRefScript datumIn -> \case
     AuthorizeHot hotCred ->
-      checkSelfPreservation inAddress inValue txInfoOutputs datumIn
+      checkSelfPreservation inAddress inValue inRefScript txInfoOutputs datumIn
         && checkMultiSig (delegationUsers datumIn) txInfoSignatories
         && checkCertificate txInfoTxCerts (TxCertAuthHotCommittee coldCred hotCred)
     ResignCold ->
-      checkSelfPreservation inAddress inValue txInfoOutputs datumIn
+      checkSelfPreservation inAddress inValue inRefScript txInfoOutputs datumIn
```
Comment (jhbertra) [inline credential-manager/src/CredentialManager/Scripts/ColdNFT.hs:46]: 2.3.2.1 - revert, not actioning

File: credential-manager/src/CredentialManager/Scripts/ColdNFT.hs [hunk 4]
```diff
@@ -37,42 +37,42 @@ coldNFTScript
...
         && checkMultiSig (membershipUsers datumIn) txInfoSignatories
         && checkCertificate txInfoTxCerts (TxCertResignColdCommittee coldCred)
     ResignDelegation user ->
-      checkContinuingTx inAddress inValue txInfoOutputs \datumOut ->
-        checkCAConservation datumIn datumOut
-          && checkNoCertificates txInfoTxCerts
-          && checkMembershipConserved datumIn datumOut
-          && checkResignation txInfoSignatories user delegationUsers datumIn datumOut
+      checkNoCertificates txInfoTxCerts
```
Comment (jhbertra) [inline credential-manager/src/CredentialManager/Scripts/ColdNFT.hs:50]: 2.3.3.3

File: credential-manager/src/CredentialManager/Scripts/ColdNFT.hs [hunk 5]
```diff
@@ -37,42 +37,42 @@ coldNFTScript
...
+        && checkContinuingTx inAddress inValue inRefScript txInfoOutputs \datumOut ->
```
Comment (jhbertra) [inline credential-manager/src/CredentialManager/Scripts/ColdNFT.hs:51]: 2.3.3.3
2.3.2.1 - revert, not actioning

File: credential-manager/src/CredentialManager/Scripts/ColdNFT.hs [hunk 6]
```diff
@@ -37,42 +37,42 @@ coldNFTScript
...
+          checkCAConservation datumIn datumOut
+            && checkMembershipConserved datumIn datumOut
+            && checkResignation txInfoSignatories user delegationUsers datumIn datumOut
     ResignMembership user ->
-      checkContinuingTx inAddress inValue txInfoOutputs \datumOut ->
-        checkCAConservation datumIn datumOut
-          && checkNoCertificates txInfoTxCerts
-          && checkDelegationConserved datumIn datumOut
-          && checkResignation txInfoSignatories user membershipUsers datumIn datumOut
+      checkNoCertificates txInfoTxCerts
```
Comment (jhbertra) [inline credential-manager/src/CredentialManager/Scripts/ColdNFT.hs:56]: 2.3.3.3

File: credential-manager/src/CredentialManager/Scripts/ColdNFT.hs [hunk 7]
```diff
@@ -37,42 +37,42 @@ coldNFTScript
...
+        && checkContinuingTx inAddress inValue inRefScript txInfoOutputs \datumOut ->
```
Comment (jhbertra) [inline credential-manager/src/CredentialManager/Scripts/ColdNFT.hs:57]: 2.3.3.3
2.3.2.1 - revert, not actioning

**Comment by jhbertra** [comment]
Appologies for not sharing this before I went on vacation - but @disassembler and I discussed what we would and would not address. Here are the notes I took for that meeting:

```
2.3.1.1 - Remediate as advised
2.3.1.2 - Remediate as advised
2.3.1.3 - No action
2.3.1.4 - Remediate as advised
2.3.1.5 - No action
2.3.2.1 - No action
2.3.2.2 - Review checks and consistently enforce == 1
2.3.2.3 - Remediate as advised
2.3.2.4 - Make documentation clearer
2.3.2.5 - Check if this is properly documented and update debug logs to be more accurate
2.3.3.1 - Remediate as required
2.3.3.2 - Review findOwnInput and consider performance ramifications
2.3.3.3 - Remediate as advised
2.3.3.4 - Remediate as advised
2.3.3.5 - Remediate as advised
2.3.3.6 - Remediate as advised
```

I have pushed some changes that revert the changes we did not want to address.

**Pull request merged**